scipy/special/tests/test_support_alternative_backends.py

```diff
@@ -18,84 +22,131 @@ def test_dispatch_to_unrecognized_library():
+    """Skip tests for specific intersections of scipy.special functions 
+    vs. backends vs. dtypes vs. devices.
+    Also suggest bespoke tweaks.
 
-    xp_assert_close(res, ref)
-
-
-@pytest.mark.fail_slow(5)
-# `reversed` is for developer convenience: test new function first = less waiting
-@pytest.mark.parametrize('f_name,n_args', reversed(array_special_func_map.items()))
-@pytest.mark.filterwarnings("ignore:invalid value encountered:RuntimeWarning:dask")
-@pytest.mark.parametrize('dtype', ['float32', 'float64'])
-@pytest.mark.parametrize('shapes', [[(0,)]*4, [tuple()]*4, [(10,)]*4,
-                                    [(10,), (11, 1), (12, 1, 1), (13, 1, 1, 1)]])
-def test_support_alternative_backends(xp, f_name, n_args, dtype, shapes):
+    Returns
+    -------
+    positive_only : bool
+        Whether you should exclusively test positive inputs.
+    dtype_np_ref : dtype
+        The dtype to use for the reference NumPy array.
+    """
     if (SCIPY_DEVICE != 'cpu'
         and is_torch(xp)
         and f_name in {'stdtr', 'stdtrit', 'betaincc', 'betainc'}
     ):
         pytest.skip(f"`{f_name}` does not have an array-agnostic implementation "
                     "and cannot delegate to PyTorch.")
+
     if is_dask(xp) and f_name == 'rel_entr':
         pytest.skip("boolean index assignment")
+
     if is_jax(xp) and f_name == "stdtrit":
         pytest.skip(f"`{f_name}` requires scipy.optimize support for immutable arrays")
+
     if is_array_api_strict(xp) and f_name == "xlogy":
         pytest.skip(f"`{f_name}` needs data-apis/array-api-strict#131 to be resolved")
 
-    shapes = shapes[:n_args]
-    f = getattr(special, f_name)
+    if 'int' in dtype and (
+        (is_torch(xp) and f_name in {'gammainc', 'gammaincc'})
+        or (is_cupy(xp) and f_name in {'stdtr', 'i0e', 'i1e'})
+        or (is_jax(xp) and f_name in {'stdtr', 'ndtr', 'ndtri', 'log_ndtr'})
+    ):
+        pytest.skip(f"`{f_name}` does not support integer types")
+
+    if ((is_jax(xp) and f_name == 'gammaincc')  # google/jax#20699
+        # gh-20972
+        or ((is_cupy(xp) or is_jax(xp) or is_torch(xp)) and f_name == 'chdtrc')):
+        positive_only = True
+    else:
+        positive_only = False
+
+    if ('int' in dtype and is_torch(xp) and xp.asarray(1.).dtype == xp.float32
+        and f_name not in {'betainc', 'betaincc', 'stdtr', 'stdtrit'}
+    ):
+        # When xp promotes int to float32, explicitly convert numpy arrays to float32
+        # to prevent them from being automatically promoted to float64 instead.
+        # To reproduce:
+        #     SCIPY_DEFAULT_DTYPE=float32 python dev.py test -b torch
+        dtype_np_ref = np.float32
+    else:
+        dtype_np_ref = getattr(np, dtype)
+
+    return positive_only, dtype_np_ref
 
+
+@pytest.mark.parametrize('f_name,n_args', array_special_func_map.items())
+@pytest.mark.filterwarnings("ignore:invalid value encountered:RuntimeWarning:dask")
```

Comment:
This could be removed if we rewrote the generic version of `xlogy` with `xpx.apply_where`, following the pattern already ubiquitously used throughout scipy.stats. However, we can't because it breaks marray: https://github.com/mdhaber/marray/issues/97